File: core/src/main/java/forpleuvoir/dhwu1a/core/common/data/OperatorData.java
package forpleuvoir.dhwu1a.core.common.data;

import com.google.gson.annotations.SerializedName;
import forpleuvoir.dhwu1a.core.user.Group;
import forpleuvoir.dhwu1a.core.user.Member;
import forpleuvoir.dhwu1a.core.user.base.Permission;

import javax.annotation.Nullable;
import java.util.Optional;
import java.util.concurrent.atomic.AtomicReference;

import static forpleuvoir.dhwu1a.core.common.ApiKey.*;

/**
 * @author forpleuvoir
 * <p>#project_name dhwu1a
 * <p>#package forpleuvoir.dhwu1a.core.common.data
 * <p>#class_name OperatorData
 * <p>#create_time 2021/7/3 22:33
 */
public class OperatorData extends UserData<Member> {
    /**
     * 操作者的群名片
     */
    @SerializedName(MEMBER_NAME)
    public final String memberName;
    /**
     * 群头衔
     */
    @SerializedName(SPECIAL_TITLE)
    public final String specialTitle;
    /**
     * 操作者在群中的权限，OWNER、ADMINISTRATOR或MEMBER
     */
    @SerializedName(PERMISSION)
    public final Permission permission;
    /**
     * 入群时间
     */
    @SerializedName(JOIN_TIMESTAMP)
    public final Integer joinTimestamp;
    /**
     * 最后发言事件
     */
    @SerializedName(LAST_SPEAK_TIMESTAMP)
    public final Integer lastSpeakTimestamp;
    /**
     * 禁言剩余时间
     */
    @SerializedName(MUTE_TIME_REMAINING)
    public final Integer muteTimeRemaining;
    /**
     * 所在群信息
     */
    @SerializedName(GROUP)
    public final GroupData group;

    public OperatorData(Long id, String memberName, String specialTitle,
                        Permission permission,
                        Integer joinTimestamp,
                        Integer lastSpeakTimestamp,
                        Integer muteTimeRemaining,
                        GroupData group
    ) {
        super(id);
        this.memberName = memberName;
        this.specialTitle = specialTitle;
        this.permission = permission;
        this.joinTimestamp = joinTimestamp;
        this.lastSpeakTimestamp = lastSpeakTimestamp;
        this.muteTimeRemaining = muteTimeRemaining;
        this.group = group;
    }

    @Nullable
    @Override
    public Member getUser() {
        AtomicReference<Member> member = new AtomicReference<>();
        Optional.ofNullable(getGroup()).ifPresent(group1 -> member.set(group1.getMember(id)));
        return member.get();
    }

    @Nullable
    public Group getGroup() {
        return group.getUser();
    }
}

File: core/src/main/kotlin/forpleuvoir/dhwu1a/core/common/data/OperatorData.kt
package forpleuvoir.dhwu1a.core.common.data

import com.google.gson.annotations.SerializedName
import forpleuvoir.dhwu1a.core.common.*
import forpleuvoir.dhwu1a.core.user.Group
import forpleuvoir.dhwu1a.core.user.Member
import forpleuvoir.dhwu1a.core.user.base.Permission
import java.util.*
import java.util.concurrent.atomic.AtomicReference

/**
 * @author forpleuvoir
 *
 * #project_name dhwu1a
 *
 * #package forpleuvoir.dhwu1a.core.common.data
 *
 * #class_name OperatorData
 *
 * #create_time 2021/7/3 22:33
 */
class OperatorData(
    id: Long?,
    /**
     * 操作者的群名片
     */
    @field:SerializedName(MEMBER_NAME) val memberName: String,
    /**
     * 群头衔
     */
    @field:SerializedName(SPECIAL_TITLE) val specialTitle: String,
    /**
     * 操作者在群中的权限，OWNER、ADMINISTRATOR或MEMBER
     */
    @field:SerializedName(PERMISSION) val permission: Permission,
    /**
     * 入群时间
     */
    @field:SerializedName(JOIN_TIMESTAMP) val joinTimestamp: Int,
    /**
     * 最后发言事件
     */
    @field:SerializedName(LAST_SPEAK_TIMESTAMP) val lastSpeakTimestamp: Int,
    /**
     * 禁言剩余时间
     */
    @field:SerializedName(MUTE_TIME_REMAINING) val muteTimeRemaining: Int,
    /**
     * 所在群信息
     */
    @field:SerializedName(GROUP) val group: GroupData
) : UserData<Member?>(id!!) {
    override val user: Member?
        get() {
            val member: AtomicReference<Member> = AtomicReference<Member>()
            Optional.ofNullable(getGroup()).ifPresent { group1: Group -> member.set(group1.getMember(id)) }
            return member.get()
        }

    fun getGroup(): Group {
        return group.user
    }
}